src/determine-testpath.js

```diff
@@ -0,0 +1,210 @@
+    return true;
+  }
+  catch (err) {
+    if (err.code !== "ENOENT") {
+      throw err;
+    }
+    return false;
+  }
+}
+
+/**
+ * Helper function that fetches the latest version of the WPT repository,
+ * restricting the checkout to META.yml files
+ */
+function fetchWPT() {
+  setupCacheFolder();
+  if (wptFolderExists()) {
+    // Pull latest commit from master branch
+    execSync("git pull origin master", { cwd: wptFolder });
+  }
+  else {
+    // Clone repo using sparse mode: the repo is huge and we're only interested
+    // in META.yml files
+    execSync("git clone https://github.com/web-platform-tests/wpt.git --depth 1 --sparse", { cwd: cacheFolder });
+    execSync("git sparse-checkout add **/META.yml", { cwd: wptFolder });
+  }
+}
+
+/**
+ * Helper function that reads "spec" entries in all META.yml files of the WPT
+ * repository.
+ *
+ * Note the function parses META.yml files as regular text files. That works
+ * well but a proper YAML parser would be needed if we need to handle things
+ * such as comments.
+ */
+async function readWptMetaFiles() {
+  async function readFolder(folder) {
+    let res = [];
+    const contents = await fs.promises.readdir(folder);
+    for (const name of contents) {
+      const filename = path.resolve(folder, name);
+      const stat = await fs.promises.stat(filename);
+      if (stat.isDirectory()) {
+        const nestedFiles = await readFolder(filename);
+        res = res.concat(nestedFiles);
+      }
+      else if (name === "META.yml") {
+        const file = await fs.promises.readFile(filename, "utf8");
+        const match = file.match(/(?:^|\n)spec: (.*)$/m);
+        if (match) {
+          res.push({
+            folder: folder.substring(wptFolder.length + 1).replace(/\\/g, "/"),
+            spec: match[1]
+          });
+        }
+      }
+    }
+    return res;
+  }
+
+  fetchWPT();
+  return await readFolder(wptFolder);
+}
+
+
+/**
+ * Returns the first item in the list found in the array, or null if none of
+ * the items exists in the array.
+ */
+function getFirstFoundInArray(paths, ...items) {
+  for (const item of items) {
+    const path = paths.find(p => p === item);
+    if (path) {
+      return path;
+    }
+  }
+  return null;
+}
+
+
+/**
+ * Exports main function that takes a list of specs as input, completes entries
+ * with a tests property when possible and returns the list.
+ *
+ * The options parameter is used to specify the GitHub API authentication token.
+ */
+module.exports = async function (specs, options) {
+  if (!specs || specs.find(spec => !spec.shortname || !spec.series || !spec.series.shortname)) {
+    throw "Invalid list of specifications passed as parameter";
+  }
+  options = options || {};
+
+  const wptFolders = await readWptMetaFiles();
+
+  function determineTestInfo(spec) {
+    const info = {
+      repository: "https://github.com/web-platform-tests/wpt"
+    };
+
+    if (spec.tests) {
+      if (spec.tests.repository) {
```

Comment:
couldn't this whole block be simplified in 
`return Object.assign({}, spec.tests);`?

Reply:
More `return Object.assign(info, spec.tests);` as we'll want to use the WPT repo by default unless otherwise specified in specs.json, but indeed, that can be simplified.

Reply:
Well, `repository` is mandatory for now under `tests` in `specs.json`, so your code would work as well in practice, but we might want to relax that rule in the future, so I'll continue to extend `info`.